agent/tasks/executetestrun.go

```diff
@@ -54,84 +75,34 @@ func (ett ExecuteTestRunTask) Run(ac *cache.AgentCache) error {
+	tr, err := ac.GetTestRun(t.TestUUID)
 	if err != nil {
-		log.Error(err)
-		return errors.New("Problem retrieving testrun from agent cache")
+		return errors.Wrapf(err, "retrieving testrun from agent cache")
 	}
 
-	log.Debugf("IMMA FIRIN MAH LAZER (for test %s) ", ett.TestUUID)
-
-	// Specify size of wait group equal to number of targets
-	wg.Add(len(tr.Targets))
+	log.Debugf("IMMA FIRIN MAH LAZER (for test %s) ", t.TestUUID)
 
-	testletPath, err := testing.GetTestletPath(tr.Testlet, ett.Config.LocalResources.OptDir)
+	testletPath, err := testing.GetTestletPath(tr.Testlet, cfg.LocalResources.OptDir)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "getting testlet path")
 	}
 
 	// Execute testlets against all targets asynchronously
-	for i := range tr.Targets {
-
-		thisTarget := tr.Targets[i]
-
-		go func() {
-
-			defer wg.Done()
-
-			log.Debugf("Full testlet command and args: '%s %s %s'", testletPath, thisTarget, tr.Args)
-			cmd := exec.Command(testletPath, thisTarget, tr.Args)
-			cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
-
-			// Stdout buffer
-			cmdOutput := &bytes.Buffer{}
-			// Attach buffer to command
-			cmd.Stdout = cmdOutput
-
-			// Execute testlet
-			cmd.Start()
-
-			done := make(chan error)
-			go func() {
-				done <- cmd.Wait()
-			}()
-
-			// This select statement will block until one of these two conditions are met:
-			// - The testlet finishes, in which case the channel "done" will be receive a value
-			// - The configured time limit is exceeded (expected for testlets running in server mode)
-			select {
-			case <-time.After(time.Duration(ett.TimeLimit) * time.Second):
-
-				if err := syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL); err != nil {
-					log.Errorf("Failed to kill %s after timeout: %s", testletPath, err)
-				} else {
-					log.Debug("Successfully killed ", testletPath)
-				}
-
-				return
-
-			case err := <-done:
-				if err != nil {
-					log.Errorf("Testlet %s completed with error '%s'", testletPath, err)
-
-					// TODO(mierdin): Handling testrun errors is on my plate as it is, and so
-					// this should get addressed properly in the future. The current approach
-					// of adding "error" to this map does nothing, as the status is tracked elsewhere
-					// gatheredData[thisTarget] = "error"
-
-					return
-				}
-				log.Debugf("Testlet %s completed without error", testletPath)
-
-			}
-
-			// Record test data
-			b := json.RawMessage(cmdOutput.Bytes())
-			gatheredData[thisTarget] = &b
-		}()
+	results := make(chan executeResult)
+	for _, target := range tr.Targets {
+		go t.executeTestRun(target, testletPath, tr.Args, results)
 	}
 
-	wg.Wait()
+	// gatheredData represents test data from this agent for all targets.
+	// Key is target name, value is JSON output from testlet for that target
+	// This is reset to a blank map every time ExecuteTestRunTask is called
+	gatheredData := map[string]*json.RawMessage{}
+	for range tr.Targets {
+		result := <-results
```

Comment:
I introduced the results channel to gather results all in the same goroutine. (It's not safe to write to a map concurrently, even if all writes are to different keys.) Since the channel is a synchronization mechanism the wait group was redundant.